kafka/kafka.sh

```diff
@@ -24,49 +28,78 @@ function update_apt_get() {
-  service zookeeper-server status || \
-      (echo 'Required zookeeper-server not running on master!' && exit 1)
-  # On master nodes, just install kafka libs but not kafka-server.
-  apt-get install -y kafka
-  exit 0
-fi
-
-# Find zookeeper list first, before attempting any installation.
-ZOOKEEPER_CLIENT_PORT=$(grep clientPort /etc/zookeeper/conf/zoo.cfg | cut -d '=' -f 2)
-ZOOKEEPER_LIST=$(grep "^server\." /etc/zookeeper/conf/zoo.cfg | \
-        cut -d '=' -f 2 | cut -d ':' -f 1 | sed "s/$/:${ZOOKEEPER_CLIENT_PORT}/" | \
-        xargs echo | sed "s/ /,/g")
-if [[ -z "${ZOOKEEPER_LIST}" ]]; then
-  # Didn't find zookeeper quorum in zoo.cfg, but possibly workers just didn't
-  # bother to populate it. Check if YARN HA is configured.
-  ZOOKEEPER_LIST=$(bdconfig get_property_value --configuration_file \
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+}
+
+function install_and_configure_kafka_server() {
+  # Find zookeeper list first, before attempting any installation.
+  local ZOOKEEPER_CLIENT_PORT
+  local ZOOKEEPER_LIST
+  readonly KAFKA_PROP_FILE='/etc/kafka/conf/server.properties'
+
+  ZOOKEEPER_CLIENT_PORT=$(grep 'clientPort' /etc/zookeeper/conf/zoo.cfg \
+    | cut -d '=' -f 2)
+  ZOOKEEPER_LIST=$(grep '^server\.' /etc/zookeeper/conf/zoo.cfg \
+    | cut -d '=' -f 2 \
+    | cut -d ':' -f 1 \
+    | sed "s/$/:${ZOOKEEPER_CLIENT_PORT}/" \
+    | xargs echo  \
+    | sed "s/ /,/g")
+
+  if [[ -z "${ZOOKEEPER_LIST}" ]]; then
+    # Didn't find zookeeper quorum in zoo.cfg, but possibly workers just didn't
+    # bother to populate it. Check if YARN HA is configured.
+    ZOOKEEPER_LIST=$(bdconfig get_property_value --configuration_file \
       /etc/hadoop/conf/yarn-site.xml \
       --name yarn.resourcemanager.zk-address 2>/dev/null)
-fi
-
-# If all attempts failed, error out.
-if [[ -z "${ZOOKEEPER_LIST}" ]]; then
-  echo 'Failed to find configured Zookeeper list; try --num-masters=3 for HA'
-  exit 1
-fi
-
-# Install Kafka from Dataproc distro.
-apt-get install -y kafka-server || dpkg -l kafka-server
-
-mkdir -p /var/lib/kafka-logs
-chown kafka:kafka -R /var/lib/kafka-logs
-
-# Note: If modified to also run brokers on master nodes, this logic for
-# generating BROKER_ID will need to be changed.
-BROKER_ID=$(hostname | sed 's/.*-w-\([0-9]\)*.*/\1/g')
-sed -i 's|log.dirs=/tmp/kafka-logs|log.dirs=/var/lib/kafka-logs|' /etc/kafka/conf/server.properties
-sed -i 's|^\(zookeeper\.connect=\).*|\1'${ZOOKEEPER_LIST}'|' /etc/kafka/conf/server.properties
-sed -i 's,^\(broker\.id=\).*,\1'${BROKER_ID}',' /etc/kafka/conf/server.properties
-echo 'delete.topic.enable = true' >> /etc/kafka/conf/server.properties
-
-# Start Kafka
-service kafka-server restart
+  fi
+
+  # If all attempts failed, error out.
+  if [[ -z "${ZOOKEEPER_LIST}" ]]; then
+    err 'Failed to find configured Zookeeper list; try --num-masters=3 for HA'
+    exit 1
+  fi
+
+  # Install Kafka from Dataproc distro.
+  apt-get install -y kafka-server \
```

Comment:
This is a change of behavior. `dpkg -l` appears to be checking if the package is already installed and allowing configuration to continue. Please keep the previous behavior or explain why this change is necessary.

Reply:
This doesn't change behaviour as if kafka-server installs successfully, then `dpkg -l` will not even run, but when kafka-server installation fails that dpkg -l would be run and if it finds already installed packaged that it will allow to continue with initialization script execution.. I assumed that it's rather unlikely that machines would already have kafka-server installed and installation will fail in the same time, but I must admin that my change modifies behaviour. I will keep that old flow, but add modified err message.